File: p2234.cpp
#include<bits/stdc++.h>
using namespace std;
#define MAX 51
int N, M, cnt, area = 1, tmp, sum;
vector<vector<int>> MAP(MAX, vector<int>(MAX, 0));
vector<vector<int>> visited(MAX, vector<int>(MAX, 0));
int dx[] = {1, 0, -1, 0};
int dy[] = {0, 1, 0, -1};
int wall[] = {4, 8, 1, 2};

void visitedClear(){
    for(int i = 0; i < N; i++) fill(visited[i].begin(), visited[i].end(), 0);
}
void print(){
    for(int i = 0 ; i < N; i++){
        for(int j = 0; j < M; j++){
            cout << visited[i][j] << " ";
        }
        cout << "\n";
    }
}
void dfs(int y, int x){
    tmp++;
    for(int i = 0; i < 4; i++){
        if(MAP[y][x] & wall[i]) continue;
        int nx = x + dx[i];
        int ny = y + dy[i];
        if(nx < 0 || ny < 0 || nx >= M || ny >= N || visited[ny][nx]) continue;
        visited[ny][nx] = visited[y][x] + 1;
        dfs(ny, nx);
    }
}
int main(){
    cin >> M >> N;
    for(int i = 0; i < N; i++){
        for(int j = 0; j < M; j++){
            cin >> MAP[i][j];
        }
    }

    for(int i = 0; i < N; i++){
        for(int j = 0; j < M; j++){
            if(visited[i][j]) continue;
            visited[i][j] = 1;
            tmp = 0;
            dfs(i, j);
            area = max(area, tmp);
            cnt++;
        }
    }

    for(int i = 0; i < N; i++){
        for(int j = 0; j < M; j++){
            for(int k = 0; k < 4; k++){
                if(MAP[i][j] & wall[k]){
                    int ny = i + dy[k];
                    int nx = j + dx[k];
                    if(nx < 0 || ny < 0 || nx >= M || ny >= N) continue;
                    visitedClear();
                    int idx;
                    if(k <= 1) idx = k + 2;
                    else idx = k - 2;
                    MAP[ny][nx] -= wall[idx];
                    MAP[i][j] -= wall[k];
                    tmp = -1;
                    dfs(i, j);
                    sum = max(sum, tmp);
                    MAP[ny][nx] += wall[idx];
                    MAP[i][j] += wall[k];
                }
            }
        }
    }

    cout << cnt << "\n";
    cout << area << "\n";
    cout << sum << "\n";
}

File: p2234.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;

public class p2234 {
    public static int cnt, mx, mx2;
    public static int N, M, tmp;
    public static int[] dy = {0, 1, 0, -1};
    public static int[] dx = {1, 0, -1, 0};
    public static int[] dir = {4, 8, 1, 2};
    public static int[][] map;
    public static boolean[][] visited;
    public static void main(String[] args) throws IOException{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        String[] input = br.readLine().split(" ");
        M = Integer.parseInt(input[0]);
        N = Integer.parseInt(input[1]);
        map = new int[N][M];
        visited = new boolean[N][M];

        for(int i = 0; i < N; i++){
            input = br.readLine().split(" ");

            for(int j = 0; j < M; j++){
                map[i][j] = Integer.parseInt(input[j]);
            }
        }

        for(int i = 0; i < N; i++){
            for(int j = 0; j < M; j++){
                if(!visited[i][j]){
                    cnt++;
                    dfs(i, j);
                    mx = Math.max(mx, tmp);
                    tmp = 0;
                }
            }
        }

        for(int y = 0; y < N; y++){
            for(int x = 0; x < M; x++){

                for(int i = 0; i < 4; i++){
                    int ny = y + dy[i];
                    int nx = x + dx[i];

                    if(ny < 0 || nx < 0 || ny >= N || nx >= M) continue;
                    if((dir[i] & map[y][x]) == 0) continue;
                    visited = new boolean[N][M];

                    map[y][x] -= dir[i];

                    dfs(y, x);
                    mx2 = Math.max(mx2, tmp);
                    tmp = 0;

                    map[y][x] += dir[i];
                }
            }
        }

        System.out.println(cnt);
        System.out.println(mx);
        System.out.println(mx2);
    }

    public static void dfs(int y, int x){
        visited[y][x] = true;
        tmp++;
        for(int i = 0; i < 4; i++){
            int ny = y + dy[i];
            int nx = x + dx[i];

            if(ny < 0 || nx < 0 || ny >= N || nx >= M || visited[ny][nx]) continue;
            if(noBarrier(dir[i], map[y][x])){
                dfs(ny, nx);
            }
        }
    }

    public static boolean noBarrier(int d, int room){
        return (d & room) == 0;
    }


}
